api-spec/filters.md

```diff
@@ -0,0 +1,133 @@
+is a simple array of WGS84 Longitude and Latitude values. They are in this order: [west, south, east, north]. 
+This query will perform an intersects operation on the geometry values of the items in the catalog. Some GeoJSON 
+objects may provide a bbox property in addition to geometry, but it should not be used for the bbox filter since
+it is an optional field in GeoJSON.
+
+example
+```
+{
+  "bbox": [-180,-90,180,90]
+}
+```
+
+The temporal query will be based on ISO 8601 and should support time ranges as well as equality. To support range
+queries, we are using a simple JSON based language. Ranges will be specified as an object with keys indicating the comparison to use.
+
+Equality is specified as `{"time": "2018-03-20T16:11:44.353Z"}`  
+Before is `{"time":{"lt":"2018-03-20T16:11:44.353Z"}}`  
+After is `{"time":{"gt":"2018-03-20T16:11:44.353Z"}}`  
+Before with Equality is `{"time":{"lte":"2018-03-20T16:11:44.353Z"}}`  
+After with Equality is `{"time":{"gte":"2018-03-20T16:11:44.353Z"}}`  
+
+Filter Extensions
+-----------------
+
+Spatial and Temporal filtering is not enough for many use cases. STAC is designed to be extensible, so we encourage additional filters to be provided as extensions.
+
+Filter extensions are intended to provide a place to specify additional filter types and behavior. Some filters may need to target vendor specific fields or use a more complex structure when defining the filter. For example, a geospatial intersection filter that supports a polygon as input.
+
+Filter extensions should be specified as an openapi fragment. This is a section of YAML that can be incorporated into the larger api definition. These fragments can then be combined into an implementation specific api definition that describes the full api.
+
+Example fragment:
+```
+  IntersectsFilter:
+    type: object
+    description: >-
+          Only returns items that intersect with the provided polygon.
+    properties:
+      intersects:
+        $ref: '#/definitions/Polygon'
+  Polygon:
+    type: object
+    properties:
+      type:
+        type: string
+        enum:
+          - polygon
+      coordinates:
+        $ref: '#/definitions/Polygon2D'
+    required:
+      - type
+      - coordinates
+  Polygon2D:
+    type: array
+    minItems: 1
+    items:
+      $ref: '#/definitions/LinearRing2D'
+  LinearRing2D:
+    type: array
+    minItems: 4
+    items:
+      $ref: '#/definitions/Point2D'
+    example: [[-104,35],[-103,35],[-103,34],[-104,34],[-104,35]]
+  Point2D:
+    description: A 2d geojson point
+    type: array
+    minimum: 2
+    maximum: 2
+    items:
+      type: number
+      format: double
+    example:
+      - -104
+      - 35
+```
+
+It is likely that there are schemas that should be used in common for types of filters that target different fields. We should define a common set of filter types that can be used in defining filters for different fields.
+- NumberRange
+- TimeRange
+- Text
+- ArrayIncludes
```

Comment:
Will this cover presence of assets? I think that'd be the other good one to think about pulling up explicitly, perhaps even put in core. Enable search by presence of 'cloud optimized geotiffs' or 'json metadata'. Or since we made it optional, by thumbnail.

Reply:
I hesitate a little on querying on the assets array. Are the types of assets strongly defined?

I was hoping that there would be enough information in the properties section so the assets don't need to be queryable. But if needed, I think a text query of ` {"assets.type": "cloud-geotiff"}` might be able to do it. It would only return records that can match that criteria. I'm not sure about all databases, but I think using the dot syntax on an array assumes an exists query on elements in the array.